File: Day147_2025/Q147_GFG.java
import java.util.*;

class Q147_GFG {
    public ArrayList<Integer> leafNodes(int[] preorder) {
        Stack<Integer> stack = new Stack<>();
        ArrayList<Integer> leaves = new ArrayList<>();
        int n = preorder.length;

        for (int i = 0, j = 1; j < n; i++, j++) {
            boolean found = false;

            if (preorder[i] > preorder[j]) {
                stack.push(preorder[i]);
            } else {
                while (!stack.isEmpty() && preorder[j] > stack.peek()) {
                    stack.pop();
                    found = true;
                }
            }

            if (found) {
                leaves.add(preorder[i]);
            }
        }

        // Add the last node as a leaf
        leaves.add(preorder[n - 1]);
        return leaves;
    }
}

File: Day147_2025/Q147_GFG.py
class Solution:
    def leafNodes(self, preorder):
        stack = []
        leaves = []
        n = len(preorder)

        for i in range(n - 1):
            found = False
            if preorder[i] > preorder[i + 1]:
                stack.append(preorder[i])
            else:
                while stack and preorder[i + 1] > stack[-1]:
                    stack.pop()
                    found = True
            if found:
                leaves.append(preorder[i])

        leaves.append(preorder[-1])  # last node is always a leaf
        return leaves
